### PR TITLE
Thread-safe implementations of MKL kernels and rerolling changes lost in previous merge

### PR DESCRIPTION
**Pull request by jbobba**
- Changes to mkl_matmul BUILD rule so it is built as part of the matmul kernel library
- Bug fixes to MKL kernels for thread-safety
- Adding a TF to MKL dimension ordering map to track dimension order of MKL tensors.
- Rerolling changes to MKL layout pass, mkl_util.h, nn_ops.cc and BUILD files that got lost with parallel merges

### CODE REVIEW

**Comment by yifeif** [comment]
@jbobba thanks for the PR. Could you resolve the conflicts? @rmlarsen mind taking a look?

**Comment by martinwicke** [comment]
I moved the if_mkl into tf_cc_test_mkl. So that part of the change shouldn't be necessary.

**Comment by jbobba** [comment]
@martinwicke do you have that change coming in from a different branch or should I remove the if_mkl in this PR?

**Comment by jbobba** [comment]
@martinwicke never mind. I checked the definition of tf_cc_test_mkl

**Review by andydavis1 — APPROVED** [review]

**Comment by jbobba** [comment]
Can we start the tests on this?

**Comment by martinwicke** [comment]
Jenkins, test this please.

File: tensorflow/core/kernels/BUILD [hunk 1]
```diff
@@ -4403,19 +4410,26 @@ tf_cc_test(
 
 if_mkl(
     tf_kernel_library(
```
Comment (martinwicke) [inline tensorflow/core/kernels/BUILD]: Could you make a tf_mkl_kernel_library and pull the if_mkl inside it? Buildozer has a bug which prevents us from inspecting/changing anything inside a top-level if, so pulling the if inside the new rule would make it a lot easier on our sync infra.

**Review by martinwicke — CHANGES_REQUESTED** [review]

**Comment by martinwicke** [comment]
One more thing related to the build.

File: tensorflow/core/kernels/BUILD (hunk 1 above)
Comment (jbobba) [inline tensorflow/core/kernels/BUILD]: Will do.

Comment (jbobba) [inline tensorflow/core/kernels/BUILD]: Can you rerun the tests?

**Comment by martinwicke** [comment]
Jenkins, test this please!

**Review by martinwicke — APPROVED** [review]

**Comment by jbobba** [comment]
Can we merge this in?

**Pull request merged**